cdap-proto/src/main/java/co/cask/cdap/proto/id/ApplicationId.java

```diff
@@ -38,6 +38,13 @@ public ApplicationId(String namespace, String application) {
 
   public ApplicationId(String namespace, String application, String version) {
     super(namespace, EntityType.APPLICATION);
+    if (application == null || application.isEmpty()) {
+      throw new IllegalArgumentException("Application ID cannot be null or empty.");
+    }
+    if (version == null) {
+      throw new IllegalArgumentException("Version cannot be null.");
```

Comment:
Should this be a NPE, and then the private helper in StreamHandler can catch both IllegalArgumentException and NPE (similar to how ConversionHelpers does it).